include/unifex/sender_for.hpp

```diff
@@ -0,0 +1,116 @@
+#include <unifex/sender_concepts.hpp>
+#include <unifex/tag_invoke.hpp>
+#include <unifex/type_traits.hpp>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex
+{
+  namespace _sf
+  {
+    template <typename Query>
+    struct _property {
+      const typename Query::value_type& operator()(const typename Query::key_type &) const noexcept {
+        return query_.value;
+      }
+      UNIFEX_NO_UNIQUE_ADDRESS Query query_;
+    };
+    template <typename... Queries>
+    struct _ctx {
+      struct type : _property<Queries>... {
+        using _property<Queries>::operator()...;
+      };
+    };
+    template <typename... Queries>
+    using _context = typename _ctx<Queries...>::type;
+
+    template <const auto& C, typename Sender, typename Context>
+    struct sender_for {
+      explicit sender_for(Sender snd, Context ctx)
+          noexcept(
+              std::is_nothrow_move_constructible_v<Sender> &&
+              std::is_nothrow_move_constructible_v<Context>)
+        : snd_((Sender&&) snd), ctx_{(Context&&) ctx} {}
+
+      // Forward all tag_invokes:
+      template (typename CPO, typename Self, typename... Args)
+        (requires same_as<sender_for, remove_cvref_t<Self>> AND
+            (!callable<const Context&, CPO>) AND
+            callable<CPO, member_t<Self, Sender>, Args...>)
+      UNIFEX_ALWAYS_INLINE
+      friend decltype(auto) tag_invoke(CPO cpo, Self&& self, Args&&... args)
+          noexcept(is_nothrow_callable_v<CPO, member_t<Self, Sender>, Args...>) {
+        return ((CPO&&) cpo)(((Self&&) self).snd_, (Args&&) args...);
+      }
+
+      // Handle custom property queries by consulting the context:
+      template (typename CPO)
+        (requires callable<const Context&, CPO>)
+      UNIFEX_ALWAYS_INLINE
+      friend decltype(auto) tag_invoke(CPO cpo, const sender_for& self) noexcept {
+        return self.ctx_(cpo);
+      }
+
+      const Sender& base() const & noexcept {
+        return snd_;
+      }
+
+      Sender&& base() && noexcept {
+        return (Sender&&) snd_;
+      }
+     private:
+      Sender snd_;
+      UNIFEX_NO_UNIQUE_ADDRESS Context ctx_;
+    };
+
+    template <auto& CPO>
+    struct _make_sender_for {
+      template (typename Sender, typename... Queries)
+        (requires sender<Sender>)
+      sender_for<CPO, remove_cvref_t<Sender>, _context<remove_cvref_t<Queries>...>>
+      operator()(Sender&& snd, Queries&&... queries) const
+          noexcept(
+              std::is_nothrow_constructible_v<remove_cvref_t<Sender>, Sender> &&
+              (std::is_nothrow_constructible_v<remove_cvref_t<Queries>, Queries> &&...)) {
+        return sender_for<CPO, remove_cvref_t<Sender>, _context<remove_cvref_t<Queries>...>>{
+            (Sender&&) snd, _context<remove_cvref_t<Queries>...>{(Queries&&) queries...}};
+      }
+    };
+  } // namespace _sf
+
+  using _sf::sender_for;
+
+  template <const auto& CPO>
+  inline constexpr _sf::_make_sender_for<CPO> make_sender_for {};
```

Comment:
I gather that the use of this `make_sender_for` helper is to wrap a sender in something that tags that sender as having been produced by a particular CPO so that you can check whether a given `co_await` expression is syntactically of the form `co_await some_cpo(...)`.

In particular, so that you can do special processing for `co_await schedule(s);` to change the current scheduler.

But this seems to come at the cost of losing some of the type information about the original sender when it is wrapped.
While the wrapper type does forward through some customisations on the sender, it only does so for customisations where the sender is the first argument, which may miss some customisations.

I wonder if there are other ways that we could tackle this requirement?

One would be defining a new tag-like function specifically for changing the current scheduler for the current coroutine.
e.g. `co_await change_current_scheduler_to(sched);`

But I guess this doesn't guard against the possibility of someone writing `co_await schedule(s);` and it just ending up being an expensive no-op which schedules on that other scheduler and then immediately reschedules back on the original context.

You could kind of guard against this somewhat by, say, deleting the `operator co_await()` on each of the schedule senders but that's not really a general solution. You could also define linters that check for the expression `co_await schedule(x);` and warn about it, but not all users will necessarily have access to those linters.

Ultimately, my concerns about the wrapper inhibiting use of customisations may not be warranted, so I'm not sure if we need to make any changes yet, but this approach does give me a little unease.

Reply:
I share your unease, but like you I don't know whether we're just being paranoid or if there is a real problem here. My feeling is, let's try it and see if it's problematic in practice. 🤷‍♂️